File: node/packages/botas/src/teams-activity.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type { CoreActivity, ChannelAccount, Conversation, Entity, Attachment } from './core-activity.js'
import type { SuggestedActions } from './suggested-actions.js'

/** Tenant information from Teams channel data. */
export interface TenantInfo {
  id?: string
  [key: string]: unknown
}

/** Channel information from Teams channel data. */
export interface ChannelInfo {
  id?: string
  name?: string
  [key: string]: unknown
}

/** Team information from Teams channel data. */
export interface TeamInfo {
  id?: string
  name?: string
  aadGroupId?: string
  [key: string]: unknown
}

/** Meeting information from Teams channel data. */
export interface MeetingInfo {
  id?: string
  [key: string]: unknown
}

/** Notification settings from Teams channel data. */
export interface NotificationInfo {
  alert?: boolean
  [key: string]: unknown
}

/** Teams-specific channel data attached to an activity. */
export interface TeamsChannelData {
  tenant?: TenantInfo
  channel?: ChannelInfo
  team?: TeamInfo
  meeting?: MeetingInfo
  notification?: NotificationInfo
  [key: string]: unknown
}

/** Teams-specific conversation with additional metadata. */
export interface TeamsConversation extends Conversation {
  conversationType?: string
  tenantId?: string
  isGroup?: boolean
  name?: string
}

/** A Teams-specific activity with strongly-typed channel data and helpers. */
export interface TeamsActivity extends CoreActivity {
  channelData?: TeamsChannelData
  timestamp?: string
  localTimestamp?: string
  locale?: string
  localTimezone?: string
  suggestedActions?: SuggestedActions
}

// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TeamsActivity = {
  /**
   * Creates a TeamsActivity from a CoreActivity.
   * Copies all fields and treats channelData as TeamsChannelData.
   *
   * @param activity - The source CoreActivity to convert.
   * @returns A shallow copy of the activity typed as {@link TeamsActivity}.
   * @throws {Error} If `activity` is falsy.
   */
  fromActivity (activity: CoreActivity): TeamsActivity {
    if (!activity) throw new Error('activity is required')
    return { ...activity } as TeamsActivity
  }
} as const

/**
 * Fluent builder for constructing outbound TeamsActivity instances.
 *
 * @example
 * ```ts
 * const reply = new TeamsActivityBuilder()
 *   .withConversationReference(incomingActivity)
 *   .withText('Hello!')
 *   .addMention(activity.from)
 *   .build()
 * ```
 */
export class TeamsActivityBuilder {
  private _activity: Partial<TeamsActivity> = { type: 'message', text: '' }

  /**
   * Copy routing fields from an incoming activity and swap from/recipient.
   *
   * @param source - The incoming activity to copy routing fields from.
   * @returns `this` for method chaining.
   */
  withConversationReference (source: CoreActivity): this {
    this._activity.serviceUrl = source.serviceUrl
    this._activity.conversation = source.conversation
    this._activity.from = source.recipient
    this._activity.recipient = source.from
    return this
  }

  /**
   * Set the activity type (default is "message").
   *
   * @param type - Activity type string.
   * @returns `this` for method chaining.
   */
  withType (type: string): this {
    this._activity.type = type
    return this
  }

  /**
   * Set the service URL.
   *
   * @param serviceUrl - Bot Framework service URL for the channel.
   * @returns `this` for method chaining.
   */
  withServiceUrl (serviceUrl: string): this {
    this._activity.serviceUrl = serviceUrl
    return this
  }

  /**
   * Set the conversation.
   *
   * @param conversation - Conversation reference.
   * @returns `this` for method chaining.
   */
  withConversation (conversation: Conversation): this {
    this._activity.conversation = conversation
    return this
  }

  /**
   * Set the sender account.
   *
   * @param from - Sender's channel account.
   * @returns `this` for method chaining.
   */
  withFrom (from: ChannelAccount): this {
    this._activity.from = from
    return this
  }

  /**
   * Set the recipient account.
   *
   * @param recipient - Recipient's channel account.
   * @returns `this` for method chaining.
   */
  withRecipient (recipient: ChannelAccount): this {
    this._activity.recipient = recipient
    return this
  }

  /**
   * Set the text content.
   *
   * @param text - Message text.
   * @returns `this` for method chaining.
   */
  withText (text: string): this {
    this._activity.text = text
    return this
  }

  /**
   * Set the Teams-specific channel data.
   *
   * @param channelData - Teams channel data (tenant, team, channel info, etc.).
   * @returns `this` for method chaining.
   */
  withChannelData (channelData?: TeamsChannelData): this {
    this._activity.channelData = channelData
    return this
  }

  /**
   * Set the suggested actions.
   *
   * @param suggestedActions - Quick-reply buttons to display below the message.
   * @returns `this` for method chaining.
   */
  withSuggestedActions (suggestedActions?: SuggestedActions): this {
    this._activity.suggestedActions = suggestedActions
    return this
  }

  /**
   * Replace the entities array.
   *
   * @param entities - New entities array, or `undefined` to clear.
   * @returns `this` for method chaining.
   */
  withEntities (entities?: Entity[]): this {
    this._activity.entities = entities
    return this
  }

  /**
   * Replace the attachments array.
   *
   * @param attachments - New attachments array, or `undefined` to clear.
   * @returns `this` for method chaining.
   */
  withAttachments (attachments?: Attachment[]): this {
    this._activity.attachments = attachments
    return this
  }

  /**
   * Set a single attachment (replaces the collection).
   *
   * @param attachment - The attachment to set.
   * @returns `this` for method chaining.
   */
  withAttachment (attachment: Attachment): this {
    this._activity.attachments = [attachment]
    return this
  }

  /**
   * Append an entity to the collection.
   *
   * @param entity - Entity to add.
   * @returns `this` for method chaining.
   */
  addEntity (entity: Entity): this {
    this._activity.entities ??= []
    this._activity.entities.push(entity)
    return this
  }

  /**
   * Append an attachment to the collection.
   *
   * @param attachment - Attachment to add.
   * @returns `this` for method chaining.
   */
  addAttachment (attachment: Attachment): this {
    this._activity.attachments ??= []
    this._activity.attachments.push(attachment)
    return this
  }

  /**
   * Creates a mention entity for the account and adds it to entities.
   * Does NOT modify the activity text — caller must include `<at>Name</at>` markup.
   *
   * @param account - The channel account to mention.
   * @param mentionText - Optional custom mention text; defaults to `<at>account.name</at>`.
   * @returns `this` for method chaining.
   * @throws {Error} If `account` is falsy.
   */
  addMention (account: ChannelAccount, mentionText?: string): this {
    if (!account) throw new Error('account is required')
    const text = mentionText ?? `<at>${account.name}</at>`
    this.addEntity({
      type: 'mention',
      mentioned: account,
      text
    })
    return this
  }

  /**
   * Parses JSON string or accepts a pre-parsed object as an Adaptive Card and appends it as an attachment.
   *
   * @param card - JSON string or pre-parsed Adaptive Card object.
   * @returns `this` for method chaining.
   * @throws {SyntaxError} If `card` is a string and not valid JSON.
   */
  addAdaptiveCardAttachment (card: string | Record<string, unknown>): this {
    const content = typeof card === 'string' ? JSON.parse(card) : structuredClone(card)
    this.addAttachment({
      contentType: 'application/vnd.microsoft.card.adaptive',
      content
    })
    return this
  }

  /**
   * Parses JSON string or accepts a pre-parsed object as an Adaptive Card and sets it as the only attachment.
   *
   * @param card - JSON string or pre-parsed Adaptive Card object.
   * @returns `this` for method chaining.
   * @throws {SyntaxError} If `card` is a string and not valid JSON.
   */
  withAdaptiveCardAttachment (card: string | Record<string, unknown>): this {
    const content = typeof card === 'string' ? JSON.parse(card) : structuredClone(card)
    this.withAttachment({
      contentType: 'application/vnd.microsoft.card.adaptive',
      content
    })
    return this
  }

  /**
   * Return the constructed TeamsActivity.
   *
   * @returns A shallow copy of the activity built so far.
   */
  build (): Partial<TeamsActivity> {
    return { ...this._activity }
  }
}

File: python/packages/botas/src/botas/teams_activity.py
"""Teams-specific activity models and builders.

Extends :class:`CoreActivity` with Teams channel data, meeting info,
suggested actions, and a fluent builder for constructing Teams messages
with mentions and Adaptive Cards.
"""

from __future__ import annotations

import copy
import json
from typing import Any

from pydantic import BaseModel, ConfigDict, model_validator
from pydantic.alias_generators import to_camel

from botas.core_activity import (
    Attachment,
    ChannelAccount,
    Conversation,
    CoreActivity,
    Entity,
)
from botas.suggested_actions import SuggestedActions


class _CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="allow",
    )


class TenantInfo(_CamelModel):
    """Microsoft 365 tenant information.

    Attributes:
        id: The tenant's unique identifier (GUID).
    """

    id: str | None = None


class ChannelInfo(_CamelModel):
    """Teams channel information.

    Attributes:
        id: Unique channel identifier.
        name: Display name of the channel.
    """

    id: str | None = None
    name: str | None = None


class TeamInfo(_CamelModel):
    """Teams team information.

    Attributes:
        id: Unique team identifier.
        name: Display name of the team.
        aad_group_id: Azure AD group ID for the team.
    """

    id: str | None = None
    name: str | None = None
    aad_group_id: str | None = None


class MeetingInfo(_CamelModel):
    """Teams meeting information.

    Attributes:
        id: Unique meeting identifier.
    """

    id: str | None = None


class NotificationInfo(_CamelModel):
    """Teams notification settings (e.g., alert flag for mobile push).

    Attributes:
        alert: When ``True``, triggers a mobile push notification.
    """

    alert: bool | None = None


class TeamsChannelData(_CamelModel):
    """Teams-specific channel data payload.

    Populated in the ``channelData`` field of a Teams activity.

    Attributes:
        tenant: Microsoft 365 tenant information.
        channel: Teams channel details.
        team: Teams team details.
        meeting: Meeting metadata (for meeting-scoped activities).
        notification: Notification settings (e.g. alert flag).
    """

    tenant: TenantInfo | None = None
    channel: ChannelInfo | None = None
    team: TeamInfo | None = None
    meeting: MeetingInfo | None = None
    notification: NotificationInfo | None = None


class TeamsConversation(Conversation):
    """Teams-specific conversation with extended metadata.

    Attributes:
        conversation_type: Type of conversation (``"personal"``, ``"channel"``, ``"groupChat"``).
        tenant_id: Microsoft 365 tenant ID.
        is_group: Whether this is a group conversation.
        name: Display name of the conversation.
    """

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="allow",
    )

    conversation_type: str | None = None
    tenant_id: str | None = None
    is_group: bool | None = None
    name: str | None = None


class TeamsActivity(CoreActivity):
    """Teams-specific activity with strongly-typed channel data and helpers.

    Extends :class:`CoreActivity` with Teams-specific fields like
    ``channel_data``, ``suggested_actions``, and timestamp fields.

    Attributes:
        channel_data: Teams-specific channel data.
        timestamp: Server-side UTC timestamp.
        local_timestamp: Client-side local timestamp.
        locale: User's locale string (e.g. ``"en-US"``).
        local_timezone: User's timezone identifier.
        suggested_actions: Quick-reply buttons.
    """

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="allow",
    )

    channel_data: TeamsChannelData | None = None
    timestamp: str | None = None
    local_timestamp: str | None = None
    locale: str | None = None
    local_timezone: str | None = None
    suggested_actions: SuggestedActions | None = None

    @model_validator(mode="before")
    @classmethod
    def _remap_from(cls, data: Any) -> Any:
        if isinstance(data, dict) and "from" in data:
            data = dict(data)
            data["from_account"] = data.pop("from")
        return data

    @staticmethod
    def from_activity(activity: CoreActivity) -> "TeamsActivity":
        """Create a TeamsActivity from a generic CoreActivity.

        Validates and re-parses the activity data to populate Teams-specific
        fields like ``channel_data``.

        Args:
            activity: A generic :class:`CoreActivity` to convert.

        Returns:
            A :class:`TeamsActivity` with Teams fields populated.

        Raises:
            ValueError: If ``activity`` is ``None``.
        """
        if activity is None:
            raise ValueError("activity is required")
        data = activity.model_dump()
        return TeamsActivity.model_validate(data)

    def add_entity(self, entity: Entity) -> None:
        """Append an entity to the activity's entities collection.

        Args:
            entity: The entity to add.
        """
        if self.entities is None:
            self.entities = []
        self.entities.append(entity)

    @staticmethod
    def create_builder() -> "TeamsActivityBuilder":
        """Return a new :class:`TeamsActivityBuilder` for fluent construction."""
        return TeamsActivityBuilder()


class TeamsActivityBuilder:
    """Fluent builder for constructing outbound TeamsActivity instances.

    Example::

        activity = (
            TeamsActivity.create_builder()
            .with_conversation_reference(incoming)
            .with_text("Hello!")
            .add_mention(user_account)
            .build()
        )
    """

    def __init__(self) -> None:
        """Initialise the builder with default values (type ``"message"``)."""
        self._type: str = "message"
        self._service_url: str = ""
        self._conversation: Conversation | None = None
        self._from_account: ChannelAccount | None = None
        self._recipient: ChannelAccount | None = None
        self._text: str = ""
        self._channel_data: TeamsChannelData | None = None
        self._suggested_actions: SuggestedActions | None = None
        self._entities: list[Entity] | None = None
        self._attachments: list[Attachment] | None = None

    def with_conversation_reference(self, source: CoreActivity) -> "TeamsActivityBuilder":
        """Copy routing fields from an incoming activity and swap from/recipient.

        Args:
            source: The incoming activity to extract routing from.

        Returns:
            The builder instance for chaining.
        """
        self._service_url = source.service_url
        self._conversation = source.conversation
        self._from_account = source.recipient
        self._recipient = source.from_account
        return self

    def with_type(self, activity_type: str) -> "TeamsActivityBuilder":
        """Set the activity type.

        Args:
            activity_type: Activity type string.

        Returns:
            The builder instance for chaining.
        """
        self._type = activity_type
        return self

    def with_service_url(self, service_url: str) -> "TeamsActivityBuilder":
        """Set the service URL.

        Args:
            service_url: Channel service endpoint URL.

        Returns:
            The builder instance for chaining.
        """
        self._service_url = service_url
        return self

    def with_conversation(self, conversation: Conversation) -> "TeamsActivityBuilder":
        """Set the conversation.

        Args:
            conversation: Target conversation.

        Returns:
            The builder instance for chaining.
        """
        self._conversation = conversation
        return self

    def with_from(self, from_account: ChannelAccount) -> "TeamsActivityBuilder":
        """Set the sender account.

        Args:
            from_account: The sender's channel account.

        Returns:
            The builder instance for chaining.
        """
        self._from_account = from_account
        return self

    def with_recipient(self, recipient: ChannelAccount) -> "TeamsActivityBuilder":
        """Set the recipient account.

        Args:
            recipient: The recipient's channel account.

        Returns:
            The builder instance for chaining.
        """
        self._recipient = recipient
        return self

    def with_text(self, text: str) -> "TeamsActivityBuilder":
        """Set the text content.

        Args:
            text: Message text.

        Returns:
            The builder instance for chaining.
        """
        self._text = text
        return self

    def with_channel_data(self, channel_data: TeamsChannelData | None) -> "TeamsActivityBuilder":
        """Set the Teams-specific channel data.

        Args:
            channel_data: Channel data payload, or ``None`` to clear.

        Returns:
            The builder instance for chaining.
        """
        self._channel_data = channel_data
        return self

    def with_suggested_actions(self, suggested_actions: SuggestedActions | None) -> "TeamsActivityBuilder":
        """Set the suggested actions.

        Args:
            suggested_actions: Quick-reply buttons, or ``None`` to clear.

        Returns:
            The builder instance for chaining.
        """
        self._suggested_actions = suggested_actions
        return self

    def with_entities(self, entities: list[Entity] | None) -> "TeamsActivityBuilder":
        """Replace the entities list.

        Args:
            entities: Entity metadata objects, or ``None`` to clear.

        Returns:
            The builder instance for chaining.
        """
        self._entities = entities
        return self

    def with_attachments(self, attachments: list[Attachment] | None) -> "TeamsActivityBuilder":
        """Replace the attachments list.

        Args:
            attachments: Attachment objects, or ``None`` to clear.

        Returns:
            The builder instance for chaining.
        """
        self._attachments = attachments
        return self

    def with_attachment(self, attachment: Attachment) -> "TeamsActivityBuilder":
        """Set a single attachment (replaces the entire collection).

        Args:
            attachment: The sole attachment.

        Returns:
            The builder instance for chaining.
        """
        self._attachments = [attachment]
        return self

    def add_entity(self, entity: Entity) -> "TeamsActivityBuilder":
        """Append an entity to the collection.

        Args:
            entity: The entity to add.

        Returns:
            The builder instance for chaining.
        """
        if self._entities is None:
            self._entities = []
        self._entities.append(entity)
        return self

    def add_attachment(self, attachment: Attachment) -> "TeamsActivityBuilder":
        """Append an attachment to the collection.

        Args:
            attachment: The attachment to add.

        Returns:
            The builder instance for chaining.
        """
        if self._attachments is None:
            self._attachments = []
        self._attachments.append(attachment)
        return self

    def add_mention(self, account: ChannelAccount, mention_text: str | None = None) -> "TeamsActivityBuilder":
        """Create a mention entity for a user.  Does NOT modify the activity text.

        You must manually include the mention text in the activity's ``text``
        field to make it visible in the chat.

        Args:
            account: The channel account to mention.
            mention_text: Custom mention markup.  Defaults to
                ``<at>{account.name}</at>``.

        Returns:
            The builder instance for chaining.

        Raises:
            ValueError: If ``account`` is ``None``.
        """
        if account is None:
            raise ValueError("account is required")
        text = mention_text or f"<at>{account.name}</at>"
        entity = Entity(type="mention", mentioned=account.model_dump(), text=text)
        return self.add_entity(entity)

    def add_adaptive_card_attachment(self, card: str | dict) -> "TeamsActivityBuilder":
        """Parse and append an Adaptive Card as an attachment.

        Args:
            card: A JSON string or pre-parsed dict representing the card.

        Returns:
            The builder instance for chaining.
        """
        content = json.loads(card) if isinstance(card, str) else copy.deepcopy(card)
        attachment = Attachment(
            content_type="application/vnd.microsoft.card.adaptive",
            content=content,
        )
        return self.add_attachment(attachment)

    def with_adaptive_card_attachment(self, card: str | dict) -> "TeamsActivityBuilder":
        """Parse and set an Adaptive Card as the sole attachment.

        Args:
            card: A JSON string or pre-parsed dict representing the card.

        Returns:
            The builder instance for chaining.
        """
        content = json.loads(card) if isinstance(card, str) else copy.deepcopy(card)
        attachment = Attachment(
            content_type="application/vnd.microsoft.card.adaptive",
            content=content,
        )
        return self.with_attachment(attachment)

    def build(self) -> TeamsActivity:
        """Build a new TeamsActivity from the current builder state.

        Returns:
            A fully constructed :class:`TeamsActivity`.
        """
        return TeamsActivity(
            type=self._type,
            service_url=self._service_url,
            conversation=self._conversation,
            from_account=self._from_account,
            recipient=self._recipient,
            text=self._text,
            channel_data=self._channel_data,
            suggested_actions=self._suggested_actions,
            entities=self._entities,
            attachments=self._attachments,
        )
